modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PageMemoryPartitionStorage.java

```diff
@@ -0,0 +1,411 @@
+        }
+
+        return clo.result();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Cursor<DataRow> scan(Predicate<SearchRow> filter) throws StorageException {
+        try {
+            IgniteCursor<TableDataRow> treeCursor = tree.find(null, null);
+
+            return new Cursor<DataRow>() {
+                @Nullable TableDataRow cur = advance();
+
+                /** {@inheritDoc} */
+                @Override
+                public void close() {
+                }
+
+                /** {@inheritDoc} */
+                @Override
+                public Iterator<DataRow> iterator() {
+                    return this;
+                }
+
+                /** {@inheritDoc} */
+                @Override
+                public boolean hasNext() {
+                    return cur != null;
+                }
+
+                /** {@inheritDoc} */
+                @Override
+                public DataRow next() {
+                    DataRow next = cur;
+
+                    if (next == null) {
+                        throw new NoSuchElementException();
+                    }
+
+                    try {
+                        cur = advance();
+                    } catch (IgniteInternalCheckedException e) {
+                        throw new StorageException("Error getting next row", e);
+                    }
+
+                    return next;
+                }
+
+                @Nullable TableDataRow advance() throws IgniteInternalCheckedException {
+                    while (treeCursor.next()) {
+                        TableDataRow dataRow = treeCursor.get();
+
+                        if (filter.test(dataRow)) {
+                            return dataRow;
+                        }
+                    }
+
+                    return null;
+                }
+            };
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Error while scanning rows", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Void> snapshot(Path snapshotPath) {
+        throw new UnsupportedOperationException("Snapshots are not supported yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void restoreSnapshot(Path snapshotPath) {
+        throw new UnsupportedOperationException("Snapshots are not supported yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void destroy() throws StorageException {
+        try {
+            tree.destroy();
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Error while destroying data", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void close() throws Exception {
+        tree.close();
+    }
+
+    private TableSearchRow wrap(SearchRow searchRow) {
+        return new TableSearchRowImpl(StorageUtils.hashCode(searchRow.key()), searchRow.keyBytes());
```

Comment:
Why do you use byte[] instead of ByteBuffers? Latter is more effective, isn't it? Can we change these classes?

Reply:
I tried to fix it.